packages/grafana-ui/src/components/Pagination/Pagination.tsx

```diff
@@ -16,21 +19,57 @@ export const Pagination: React.FC<Props> = ({ currentPage, numberOfPages, onNavi
   const styles = getStyles();
   const pages = [...new Array(numberOfPages).keys()];
 
+  const condensePages = numberOfPages >= PAGE_LENGTH_TO_CONDENSE;
+  const getListItem = (page: number, variant: 'primary' | 'secondary') => (
+    <li key={page} className={styles.item}>
+      <Button size="sm" variant={variant} onClick={() => onNavigate(page)}>
+        {page}
+      </Button>
+    </li>
+  );
+
   return (
     <div className={styles.container}>
       <ol>
-        {pages.map(pageIndex => {
+        <li className={styles.item}>
+          <Button
+            size="sm"
+            variant="secondary"
+            onClick={() => onNavigate(currentPage - 1)}
+            disabled={currentPage === 1}
+          >
+            <Icon name="angle-left" />
+          </Button>
+        </li>
+        {pages.reduce<JSX.Element[]>((pagesToRender, pageIndex) => {
           const page = pageIndex + 1;
           const variant: ButtonVariant = page === currentPage ? 'primary' : 'secondary';
 
-          return (
-            <li key={page} className={styles.item}>
-              <Button size="sm" variant={variant} onClick={() => onNavigate(page)}>
-                {page}
-              </Button>
-            </li>
-          );
-        })}
+          if (condensePages) {
+            if (page === 1 || page === numberOfPages || (page >= currentPage - 2 && page <= currentPage + 2)) {
+              pagesToRender.push(getListItem(page, variant));
+            } else if (page === currentPage - 3 || page === currentPage + 3) {
+              pagesToRender.push(
+                <li key={page} className={styles.item}>
+                  <Icon className={styles.ellipsis} name="ellipsis-v" />
+                </li>
+              );
+            }
```

Comment:
I think this logic is hard to understand and reason about without comments and or tests. Maybe you could add tests or comments so other can understand the magic numbers here without debugging the code?

Reply:
it's about to get even more complicated so I will definitely add some helpful comments and see about adding tests :)

Reply:
Thank you 🙏